File: admin/src/api/mpPoster.ts
import { fetchJSON, withToken } from '@/api/http'

export type MpPosterStatus = 'draft' | 'published' | 'offline'

export type MpPosterItem = {
  id: number
  title: string
  poster_url: string
  poster_width: number
  poster_height: number
  poster_ratio: number
  status: MpPosterStatus
  weight: number
  content_format: string
  content_text?: string
  content_nodes?: any[]
  published_at?: string
  created_at: string
  updated_at: string
}

export type MpPosterListResponse = {
  ok: boolean
  error?: string
  page: number
  page_size: number
  total: number
  items: MpPosterItem[]
}

export type MpPosterDetailResponse = {
  ok: boolean
  error?: string
  item: MpPosterItem
}

export type MpPosterUploadImageResponse = {
  ok: boolean
  error?: string
  url: string
  mime: string
  bytes: number
  width: number
  height: number
}

export async function fetchPosterList(params: { page?: number; pageSize?: number; status?: string }) {
  const qs = new URLSearchParams()
  qs.set('page', String(params.page || 1))
  qs.set('page_size', String(params.pageSize || 20))
  if (params.status) qs.set('status', params.status)
  const url = withToken(`/api/v1/admin/mp/posters?${qs.toString()}`)
  const res = await fetchJSON<MpPosterListResponse>(url)
  if (!res.ok) throw new Error(res.error || 'backend_error')
  return res
}

export async function fetchPosterDetail(id: number) {
  const url = withToken(`/api/v1/admin/mp/posters/${encodeURIComponent(String(id))}`)
  const res = await fetchJSON<MpPosterDetailResponse>(url)
  if (!res.ok) throw new Error(res.error || 'backend_error')
  return res
}

export async function createPoster(params: {
  title: string
  poster_url?: string
  poster_width?: number
  poster_height?: number
  poster_ratio?: number
  status?: MpPosterStatus
  weight?: number
  content_format?: string
  content_text?: string
  content_nodes?: any
  published_at?: string | null
}) {
  const url = withToken('/api/v1/admin/mp/posters')
  const res = await fetchJSON<MpPosterDetailResponse>(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  })
  if (!res.ok) throw new Error(res.error || 'backend_error')
  return res
}

export async function updatePoster(
  id: number,
  params: Partial<{
    title: string
    poster_url: string
    poster_width: number
    poster_height: number
    poster_ratio: number
    status: MpPosterStatus
    weight: number
    content_format: string
    content_text: string
    content_nodes: any
    published_at: string | null
  }>,
) {
  const url = withToken(`/api/v1/admin/mp/posters/${encodeURIComponent(String(id))}`)
  const res = await fetchJSON<MpPosterDetailResponse>(url, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  })
  if (!res.ok) throw new Error(res.error || 'backend_error')
  return res
}

export async function deletePoster(id: number) {
  const url = withToken(`/api/v1/admin/mp/posters/${encodeURIComponent(String(id))}`)
  const res = await fetchJSON<{ ok: boolean; error?: string }>(url, {
    method: 'DELETE',
  })
  if (!res.ok) throw new Error(res.error || 'backend_error')
  return res
}

export async function uploadPosterImage(file: File) {
  const url = withToken('/api/v1/admin/mp/poster/image')
  const token = (localStorage.getItem('f1ink_admin_token') || '').trim()
  const formData = new FormData()
  formData.append('image', file)
  if (token) formData.append('token', token)
  const extraHeaders: Record<string, string> = {}
  if (token) extraHeaders['Authorization'] = `Bearer ${token}`
  const res = await fetchJSON<MpPosterUploadImageResponse>(url, {
    method: 'POST',
    headers: extraHeaders,
    body: formData,
  })
  if (!res.ok) throw new Error(res.error || 'backend_error')
  return res
}

File: backend/internal/model/mp_poster.go
package model

import (
	"time"
)

type MpPosterStatus string

const (
	MpPosterStatusDraft     MpPosterStatus = "draft"
	MpPosterStatusPublished MpPosterStatus = "published"
	MpPosterStatusOffline   MpPosterStatus = "offline"
)

type MpPoster struct {
	ID              int64          `json:"id" gorm:"primaryKey;autoIncrement"`
	Title           string         `json:"title" gorm:"type:varchar(255);not null"`
	PosterURL       string         `json:"poster_url" gorm:"type:varchar(512);not null;default:''"`
	PosterWidth     int            `json:"poster_width" gorm:"type:int;not null;default:0"`
	PosterHeight    int            `json:"poster_height" gorm:"type:int;not null;default:0"`
	PosterRatio     float64        `json:"poster_ratio" gorm:"type:double;not null;default:0"`
	Status          MpPosterStatus `json:"status" gorm:"type:varchar(32);not null;default:draft;index"`
	Weight          int            `json:"weight" gorm:"type:int;not null;default:0"`
	ContentFormat   string         `json:"content_format" gorm:"type:varchar(32);not null;default:'RICH_TEXT_NODES'"`
	ContentText     string         `json:"content_text,omitempty" gorm:"type:mediumtext"`
	ContentNodes    string         `json:"content_nodes,omitempty" gorm:"type:json"`
	PublishedAt     *time.Time     `json:"published_at"`
	CreatedAt       time.Time      `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt       time.Time      `json:"updated_at" gorm:"autoUpdateTime"`
}

func (MpPoster) TableName() string {
	return "mp_posters"
}

type MpPosterListResponse struct {
	Ok       bool       `json:"ok"`
	Page     int        `json:"page"`
	PageSize int        `json:"page_size"`
	Total    int64      `json:"total"`
	Items    []MpPoster `json:"items"`
}

type MpPosterDetailResponse struct {
	Ok   bool     `json:"ok"`
	Item MpPoster `json:"item"`
}

type MpPosterCreateRequest struct {
	Title        string         `json:"title" binding:"required,max=255"`
	PosterURL    string         `json:"poster_url"`
	PosterWidth  int            `json:"poster_width"`
	PosterHeight int            `json:"poster_height"`
	PosterRatio  float64        `json:"poster_ratio"`
	Status       MpPosterStatus `json:"status"`
	Weight       int            `json:"weight"`
	ContentFormat string        `json:"content_format"`
	ContentText  string         `json:"content_text"`
	ContentNodes any            `json:"content_nodes"`
	PublishedAt  *time.Time     `json:"published_at"`
}

type MpPosterUpdateRequest struct {
	Title        *string         `json:"title" binding:"omitempty,max=255"`
	PosterURL    *string         `json:"poster_url"`
	PosterWidth  *int            `json:"poster_width"`
	PosterHeight *int            `json:"poster_height"`
	PosterRatio  *float64        `json:"poster_ratio"`
	Status       *MpPosterStatus `json:"status"`
	Weight       *int            `json:"weight"`
	ContentFormat *string        `json:"content_format"`
	ContentText  *string         `json:"content_text"`
	ContentNodes *any            `json:"content_nodes"`
	PublishedAt  **time.Time     `json:"published_at"`
}

type MpPosterUploadImageResponse struct {
	Ok      bool   `json:"ok"`
	URL     string `json:"url"`
	Mime    string `json:"mime"`
	Bytes   int64  `json:"bytes"`
	Width   int    `json:"width"`
	Height  int    `json:"height"`
}

type MpMpPosterResponse struct {
	Ok       bool      `json:"ok"`
	BaseURL  string    `json:"base_url"`
	Item     *MpPoster `json:"item,omitempty"`
}
